language/move-prover/stackless-bytecode-generator/src/stackless_bytecode.rs

```diff
@@ -1,19 +1,81 @@
 // Copyright (c) The Libra Core Contributors
 // SPDX-License-Identifier: Apache-2.0
 
-use vm::file_format::{
-    AddressPoolIndex, ByteArrayPoolIndex, CodeOffset, FunctionHandleIndex, SignatureIndex,
-    StructDefinitionIndex,
-};
+use num::BigUint;
+use spec_lang::ty::Type;
+use vm::file_format::{CodeOffset, FunctionHandleIndex, StructDefinitionIndex};
 
 pub type TempIndex = usize;
 
+/// The kind of an assignment in the bytecode.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum AssignKind {
+    /// The assign copies the lhs value.
+    Copy,
+    /// The assign moves the lhs value.
+    Move,
+    /// The assign stores the lhs value.
+    // TODO: figure out why we can't treat this as either copy or move. The lifetime analysis
+    // currently makes a difference of this case. It originates from stack code where Copy
+    // and Move push on the stack and Store pops.
+    Store,
+}
+
+/// A constant value.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Constant {
+    Bool(bool),
+    U8(u8),
+    U64(u64),
+    U128(u128),
+    Address(BigUint),
+    ByteArray(Vec<u8>),
+    TxnSenderAddress,
+}
+
+/// A unary operation.
+#[derive(Debug, Clone, PartialEq, Eq, Copy)]
+pub enum UnaryOp {
+    CastU8,
+    CastU64,
+    CastU128,
+    Not,
+}
+
+/// A binary operation.
+#[derive(Debug, Clone, PartialEq, Eq, Copy)]
+pub enum BinaryOp {
+    Add,
+    Sub,
+    Mul,
+    Div,
+    Mod,
+    BitOr,
+    BitAnd,
+    Xor,
+    Shl,
+    Shr,
+    Lt,
+    Gt,
+    Le,
+    Ge,
+    Or,
+    And,
+    Eq,
+    Neq,
+}
+
+/// A branch condition.
+#[derive(Debug, Clone, PartialEq, Eq, Copy)]
+pub enum BranchCond {
+    Always,
+    True(TempIndex),
+    False(TempIndex),
+}
+
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub enum StacklessBytecode {
```

Comment:
We could rename StacklessBytecode -> Bytecode.

Reply:
Done